auth/authentication_options.hh

```diff
@@ -53,8 +57,22 @@ struct authentication_options final {
     std::optional<custom_options> options;
 };
 
-inline bool any_authentication_options(const authentication_options& aos) noexcept {
-    return aos.password || aos.options;
+/// Used exclusively for restoring roles.
+struct authentication_with_salted_hash_options final {
+    sstring salted_hash;
+    std::optional<custom_options> options;
+};
+
```

Comment:
I believe this could have been made simpler with this change to authentiation_options {

```c++
struct password_option {
    sstring password;
};

struct salted_hash_option {
    sstring salted_hash;
};

struct authentication_options final {
    std::optional<std::variant<password_option, salted_hash_option>> credential;
    std::optional<custom_options> options;
};
```

I believe this expresses the choices with less nesting. But it's okay to keep the current code.